File: extensions/daemon.mjs
#!/usr/bin/env node
/**
 * pi-cron daemon — self-contained scheduler for pi-cron jobs.
 *
 * Spawned detached by the extension on first cron_add (ensureDaemon).
 * Scans $PI_CRON_DIR/jobs every 30s; due jobs run `pi -p "<prompt>"`
 * as a subprocess with the env snapshot captured at schedule time,
 * then post the result via the Telegram Bot API (if TG_BOT_TOKEN +
 * TG_CHAT were captured and job.silent isn't set).
 *
 * No dependencies beyond node + a `pi` binary on PATH.
 * Lockfile: <dir>/daemon.pid — one daemon per job dir.
 */

import {
	existsSync,
	mkdirSync,
	readFileSync,
	readdirSync,
	writeFileSync,
	unlinkSync,
	appendFileSync,
} from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";
import { spawn } from "node:child_process";

const DIR =
	process.env.PI_CRON_DIR ??
	join(homedir(), ".local/state/telegram-agent/cron");
const JOBS = join(DIR, "jobs");
const LOG = join(DIR, "daemon.log");
const PIDFILE = join(DIR, "daemon.pid");
const CATCHUP_GRACE_S = 900;
const SCAN_MS = 30_000;
const RUN_TIMEOUT_MS = 15 * 60 * 1000;

const running = new Set();

function log(...args) {
	try {
		appendFileSync(
			LOG,
			`${new Date().toISOString()} ${args.join(" ")}\n`,
		);
	} catch {}
}

// ---------- minimal cron parser (5-field, dom/dow OR rule) ----------

function fieldMatches(field, val) {
	for (const part of field.split(",")) {
		const p = part.trim();
		if (p === "*") return true;
		if (p.startsWith("*/")) {
			const step = parseInt(p.slice(2));
			if (step && val % step === 0) return true;
			continue;
		}
		if (p.includes("-")) {
			const [a, b] = p.split("-").map(Number);
			if (val >= a && val <= b) return true;
			continue;
		}
		if (parseInt(p) === val) return true;
	}
	return false;
}

const MONTHS = { jan:1,feb:2,mar:3,apr:4,may:5,jun:6,jul:7,aug:8,sep:9,oct:10,nov:11,dec:12 };
const DOWS = { sun:0,mon:1,tue:2,wed:3,thu:4,fri:5,sat:6 };

function normField(f, names) {
	return f
		.split(",")
		.map((p) => {
			const m = p.match(/^([a-z]{3})(?:-([a-z]{3}))?$/i);
			if (m && names) {
				const a = names[m[1].toLowerCase()];
				const b = m[2] ? names[m[2].toLowerCase()] : undefined;
				if (a !== undefined)
					return b !== undefined ? `${a}-${b}` : String(a);
			}
			return p;
		})
		.join(",");
}

function cronNext(expr, afterTs) {
	const f = expr.trim().split(/\s+/);
	if (f.length !== 5) return null;
	f[3] = normField(f[3], MONTHS);
	f[4] = normField(f[4], DOWS);
	const domStar = f[2] === "*";
	const dowStar = f[4] === "*";
	let t = new Date(afterTs * 1000);
	t.setSeconds(0, 0);
	t = new Date(t.getTime() + 60_000);
	for (let i = 0; i < 366 * 24 * 60; i++) {
		const dow = t.getDay();
		const domOk = fieldMatches(f[2], t.getDate());
		const dowOk = fieldMatches(f[4], dow);
		// cron rule: when BOTH dom and dow are restricted → OR; else AND
		const dayOk =
			!domStar && !dowStar ? domOk || dowOk : domOk && dowOk;
		if (
			fieldMatches(f[0], t.getMinutes()) &&
			fieldMatches(f[1], t.getHours()) &&
			dayOk &&
			fieldMatches(f[3], t.getMonth() + 1)
		) {
			return Math.floor(t.getTime() / 1000);
		}
		t = new Date(t.getTime() + 60_000);
	}
	return null;
}

function nextRun(spec, after = Math.floor(Date.now() / 1000)) {
	const [kind, arg] = spec.split(/:(.*)/s);
	if (kind === "every") {
		const m = parseInt(arg);
		return m > 0 ? after + m * 60 : null;
	}
	if (kind === "daily") {
		const [h, m] = arg.split(":").map(Number);
		const d = new Date();
		d.setHours(h, m, 0, 0);
		let ts = Math.floor(d.getTime() / 1000);
		if (ts <= after) ts += 86400;
		return ts;
	}
	if (kind === "once") {
		const ts = parseInt(arg);
		return ts > after - CATCHUP_GRACE_S ? ts : null;
	}
	if (kind === "cron") return cronNext(arg, after);
	return null;
}

// ---------- job firing ----------

async function postTelegram(env, text) {
	const tok = env.TG_BOT_TOKEN;
	const chat = env.TG_CHAT;
	if (!tok || !chat) return;
	const body = {
		chat_id: parseInt(chat),
		text: text.slice(0, 4000),
		disable_notification: false,
	};
	if (env.TG_THREAD) body.message_thread_id = parseInt(env.TG_THREAD);
	try {
		await fetch(`https://api.telegram.org/bot${tok}/sendMessage`, {
			method: "POST",
			headers: { "content-type": "application/json" },
			body: JSON.stringify(body),
		});
	} catch (e) {
		log("telegram post failed:", e.message);
	}
}

function runPi(job) {
	return new Promise((resolve) => {
		const args = ["-p", job.prompt];
		const proc = spawn("pi", args, {
			cwd: job.cwd || homedir(),
			env: { ...process.env, ...(job.env || {}) },
			timeout: RUN_TIMEOUT_MS,
		});
		let out = "";
		proc.stdout.on("data", (d) => (out += d));
		proc.stderr.on("data", (d) => log("pi stderr:", String(d).slice(0, 300)));
		proc.on("close", (code) => resolve({ code, out: out.trim() }));
		proc.on("error", (e) => resolve({ code: -1, out: "", err: e.message }));
	});
}

async function fire(job, file) {
	log("firing", job.id, job.spec, `soul=${job.soul}`);
	const { code, out, err } = await runPi(job);
	job.runs = (job.runs || 0) + 1;
	job.last_run = Math.floor(Date.now() / 1000);
	if (code !== 0) {
		job.fails = (job.fails || 0) + 1;
		log("job failed:", job.id, err || `exit ${code}`);
	}
	// post result unless silent or 'none'
	const text = (out || "").trim();
	if (!job.silent && text && !/^(none|-)$/i.test(text)) {
		await postTelegram(job.env || {}, text);
	}
	// reschedule or delete
	if (job.spec.startsWith("once:")) {
		unlinkSync(file);
		return;
	}
	const nxt = nextRun(job.spec);
	if (nxt === null) {
		unlinkSync(file);
		return;
	}
	job.next_run = nxt;
	writeFileSync(file, JSON.stringify(job, null, 2));
}

// ---------- main loop ----------

function scan() {
	if (!existsSync(JOBS)) return;
	const now = Math.floor(Date.now() / 1000);
	for (const name of readdirSync(JOBS)) {
		if (!name.endsWith(".json")) continue;
		const file = join(JOBS, name);
		let job;
		try {
			job = JSON.parse(readFileSync(file, "utf-8"));
		} catch {
			unlinkSync(file);
			continue;
		}
		const jid = job.id || name;
		if (job.paused || running.has(jid)) continue;
		let nxt = job.next_run || 0;
		if (nxt <= 0) {
			nxt = nextRun(job.spec) ?? now + 3600;
			job.next_run = nxt;
			writeFileSync(file, JSON.stringify(job));
			continue;
		}
		if (nxt <= now) {
			if (
				now - nxt > CATCHUP_GRACE_S &&
				!job.spec.startsWith("once:")
			) {
				job.next_run = nextRun(job.spec, now) ?? now + 3600;
				writeFileSync(file, JSON.stringify(job));
				continue;
			}
			running.add(jid);
			fire(job, file).finally(() => running.delete(jid));
		}
	}
}

// single-instance guard
mkdirSync(DIR, { recursive: true });
if (existsSync(PIDFILE)) {
	try {
		const pid = parseInt(readFileSync(PIDFILE, "utf-8"));
		process.kill(pid, 0);
		log("already running as pid", pid, "— exiting");
		process.exit(0);
	} catch {
		/* stale pidfile */
	}
}
writeFileSync(PIDFILE, String(process.pid));
process.on("exit", () => {
	try {
		unlinkSync(PIDFILE);
	} catch {}
});
log("pi-cron daemon up, pid", process.pid, "dir", JOBS);
scan();
setInterval(scan, SCAN_MS);

File: extensions/daemon.py
#!/usr/bin/env python3
"""pi-cron daemon — self-contained scheduler for pi-cron jobs.

Spawned by the extension on first cron_add (or via `make install`
→ systemd user unit). Scans $PI_CRON_DIR/jobs every 30s; due jobs run
`pi -p "<prompt>"` with the env snapshot captured at schedule time,
then post the result via the Telegram Bot API when TG_BOT_TOKEN +
TG_CHAT were captured and the job isn't silent.

Zero deps beyond python3 + a `pi` binary on PATH.
Lockfile: <dir>/daemon.pid — one daemon per job dir.
"""

from __future__ import annotations

import json
import os
import signal
import threading
import subprocess
import sys
import time
import urllib.request
import uuid
from datetime import datetime, timedelta
from pathlib import Path

DIR = Path(os.environ.get("PI_CRON_DIR") or
           Path.home() / ".local/state/pi-cron")
JOBS = DIR / "jobs"
LOG = DIR / "daemon.log"
PIDFILE = DIR / "daemon.pid"
CATCHUP_GRACE_S = 900
SCAN_S = 30
RUN_TIMEOUT_S = 15 * 60

running: set[str] = set()
sem = threading.BoundedSemaphore(4)  # max concurrent pi runs


def _atomic_write(path: Path, job: dict) -> None:
    tmp = path.with_suffix(".tmp")
    tmp.write_text(json.dumps(job))
    tmp.replace(path)


def _rotate_log() -> None:
    try:
        if LOG.exists() and LOG.stat().st_size > 512 * 1024:
            LOG.write_bytes(LOG.read_bytes()[-256 * 1024:])
    except OSError:
        pass


def log(*args):
    try:
        with open(LOG, "a") as f:
            f.write(f"{datetime.now().isoformat()} {' '.join(map(str, args))}\n")
    except OSError:
        pass


# ---------- minimal cron parser (5-field, dom/dow OR rule) ----------

_MONTHS = {m: i + 1 for i, m in enumerate(
    "jan feb mar apr may jun jul aug sep oct nov dec".split())}
_DOWS = {d: i for i, d in enumerate("sun mon tue wed thu fri sat".split())}


def _field_ok(field: str, val: int) -> bool:
    for part in field.split(","):
        p = part.strip()
        if p == "*":
            return True
        if p.startswith("*/"):
            step = int(p[2:] or 0)
            if step and val % step == 0:
                return True
            continue
        if "-" in p:
            a, b = (int(x) for x in p.split("-", 1))
            if a <= val <= b:
                return True
            continue
        try:
            if int(p) == val:
                return True
        except ValueError:
            pass
    return False


def _names(field: str, table: dict[str, int]) -> str:
    out = []
    for p in field.split(","):
        m = p.strip().lower()
        if "-" in m:
            a, _, b = m.partition("-")
            if a in table and b in table:
                out.append(f"{table[a]}-{table[b]}")
                continue
        out.append(str(table[m]) if m in table else p)
    return ",".join(out)


def _split_tz(arg: str) -> tuple[str, object | None]:
    """'<expr>@<IANA zone>' — daily:09:00@Asia/Tehran, cron:0 9 * * 1-5@UTC.
    Returns (expr, tzinfo|None)."""
    if "@" not in arg:
        return arg, None
    expr, _, zone = arg.partition("@")
    try:
        from zoneinfo import ZoneInfo

        return expr, ZoneInfo(zone)
    except Exception:
        log("bad timezone in spec, using host local:", arg)
        return expr, None


def cron_next(expr: str, after: int, tz=None) -> int | None:
    f = expr.split()
    if len(f) != 5:
        return None
    f[3] = _names(f[3], _MONTHS)
    f[4] = _names(f[4], _DOWS)
    dom_star, dow_star = f[2] == "*", f[4] == "*"
    t = datetime.fromtimestamp(after, tz=tz).replace(second=0, microsecond=0)
    t += timedelta(minutes=1)
    for _ in range(366 * 24 * 60):
        dom_ok = _field_ok(f[2], t.day)
        dow_ok = _field_ok(f[4], (t.weekday() + 1) % 7)  # Sun=0
        day_ok = (dom_ok or dow_ok) if not dom_star and not dow_star \
            else dom_ok and dow_ok
        if (_field_ok(f[0], t.minute) and _field_ok(f[1], t.hour)
                and day_ok and _field_ok(f[3], t.month)):
            return int(t.timestamp())
        t += timedelta(minutes=1)
    return None


def next_run(spec: str, after: int | None = None) -> int | None:
    now = after or int(time.time())
    kind, _, arg = spec.partition(":")
    if kind == "every":
        if arg.endswith("s"):
            s = int(arg[:-1] or 0)
            return now + s if s > 0 else None
        m = int(arg or 0)
        return now + m * 60 if m > 0 else None
    if kind == "daily":
        arg, tz = _split_tz(arg)
        try:
            h, m = (int(x) for x in arg.split(":"))
        except ValueError:
            return None
        d = datetime.now(tz=tz).replace(hour=h, minute=m, second=0, microsecond=0)
        if int(d.timestamp()) <= now:
            d += timedelta(days=1)
        return int(d.timestamp())
    if kind == "once":
        ts = int(arg or 0)
        return ts if ts > now - CATCHUP_GRACE_S else None
    if kind == "cron":
        arg, tz = _split_tz(arg)
        return cron_next(arg, now, tz)
    return None


# ---------- firing ----------


def deliver(env: dict, soul: str, text: str) -> None:
    """Post the run result — prefer the team mailbox (host delivers as
    the soul and journals it); fall back to direct Bot API."""
    team_dir = env.get("PI_TEAM_DIR")
    if team_dir:
        try:
            req_dir = Path(team_dir) / "requests"
            req_dir.mkdir(parents=True, exist_ok=True)
            rid = uuid.uuid4().hex
            (req_dir / f"{rid}.json").write_text(json.dumps({
                "id": rid, "from": soul, "to": "host", "kind": "say",
                "text": text[:4000],
                "chat": env.get("PI_TEAM_CHAT") or env.get("TG_CHAT"),
                "thread": env.get("PI_TEAM_THREAD"),
                "at": int(time.time() * 1000),
            }))
            return
        except Exception as e:
            log("mailbox write failed, direct post:", e)
    tok, chat = env.get("TG_BOT_TOKEN"), env.get("TG_CHAT")
    if not tok or not chat:
        return
    body = {"chat_id": int(chat), "text": text[:4000]}
    if env.get("TG_THREAD"):
        body["message_thread_id"] = int(env["TG_THREAD"])
    try:
        req = urllib.request.Request(
            f"https://api.telegram.org/bot{tok}/sendMessage",
            data=json.dumps(body).encode(),
            headers={"content-type": "application/json"},
        )
        urllib.request.urlopen(req, timeout=15)
    except Exception as e:
        log("telegram post failed:", e)


def _final_text(stdout: str) -> str:
    """Last assistant text from pi's --mode json event stream."""
    text = ""
    for line in stdout.splitlines():
        try:
            ev = json.loads(line)
        except ValueError:
            continue
        msg = ev.get("message") or {}
        if (ev.get("type") == "message_end"
                and msg.get("role") == "assistant"):
            parts = [p.get("text", "") for p in msg.get("content", [])
                     if isinstance(p, dict) and p.get("type") == "text"]
            if parts:
                text = "".join(parts)
    return text


def fire(job: dict, path: Path) -> None:
    log("firing", job["id"], job["spec"], f"soul={job.get('soul')}")
    env = {**os.environ, **(job.get("env") or {})}
    # jobs snapshot GH_TOKEN at create-time — it expires within the
    # hour. The gh shim re-mints fresh per call; drop the stale one.
    env.pop("GH_TOKEN", None)
    repo = env.get("PI_TEAM_REPO")
    if repo:
        env["PATH"] = f"{repo}/tools/bin:{env.get('PATH', '')}"
    args = ["pi", "-p", job["prompt"], "--mode", "json", "--no-session"]
    if job.get("tools"):
        args += ["--tools", job["tools"]]
    # extension (monitoring/team tools) — job may pin one, else the
    # default from env so cron runs get the same tools as chats
    ext = job.get("extension") or os.environ.get("PI_EXTENSION", "")
    if ext:
        args += ["--extension", ext]
    try:
        r = subprocess.run(
            args,
            cwd=job.get("cwd") or str(Path.home()),
            env=env, capture_output=True, text=True,
            timeout=RUN_TIMEOUT_S,
        )
        out = _final_text(r.stdout or "") or (r.stdout or "").strip()
        job["last_status"] = "ok" if r.returncode == 0 else "failed"
        if r.returncode != 0:
            job["fails"] = int(job.get("fails") or 0) + 1
            log("job failed:", job["id"], "exit", r.returncode,
                (r.stderr or "")[:200])
    except subprocess.TimeoutExpired:
        out, job["fails"] = "", int(job.get("fails") or 0) + 1
        job["last_status"] = "timeout"
        log("job timed out:", job["id"])
    except Exception as e:
        out, job["fails"] = "", int(job.get("fails") or 0) + 1
        job["last_status"] = "error"
        log("job error:", job["id"], e)

    job["runs"] = int(job.get("runs") or 0) + 1
    job["last_run"] = int(time.time())
    job["last_answer"] = (out or "")[:200]
    if not job.get("silent") and out and out.lower() not in ("none", "-"):
        deliver(job.get("env") or {}, job.get("soul", ""), out)

    times = int(job.get("times") or 0)
    if job["spec"].startswith("once:") or (times and job["runs"] >= times):
        path.unlink(missing_ok=True)
        log("job done:", job["id"], f"({job['runs']} runs)")
        return
    nxt = next_run(job["spec"])
    if nxt is None:
        path.unlink(missing_ok=True)
        return
    job["next_run"] = nxt
    _atomic_write(path, job)


# ---------- loop ----------


def _pid_alive(pid: int) -> bool:
    try:
        os.kill(pid, 0)
        return True
    except OSError:
        return False


def main() -> None:
    JOBS.mkdir(parents=True, exist_ok=True)
    if PIDFILE.exists():
        try:
            if _pid_alive(int(PIDFILE.read_text())):
                log("already running — exit")
                return
        except (ValueError, OSError):
            pass
    PIDFILE.write_text(str(os.getpid()))
    log("pi-cron daemon up, pid", os.getpid(), "dir", JOBS)

    def _bye(*_):
        PIDFILE.unlink(missing_ok=True)
        sys.exit(0)

    signal.signal(signal.SIGTERM, _bye)
    signal.signal(signal.SIGINT, _bye)

    while True:
        try:
            _rotate_log()
            (DIR / "heartbeat").write_text(str(int(time.time())))
            now = int(time.time())
            for f in sorted(JOBS.glob("*.json")):
                try:
                    job = json.loads(f.read_text())
                except (OSError, ValueError):
                    f.unlink(missing_ok=True)
                    continue
                jid = job.get("id", f.stem)
                if jid in running or job.get("paused"):
                    continue
                nxt = int(job.get("next_run") or 0)
                if nxt <= 0:
                    job["next_run"] = next_run(job["spec"]) or now + 3600
                    _atomic_write(f, job)
                    continue
                if nxt <= now:
                    if (now - nxt > CATCHUP_GRACE_S
                            and not job["spec"].startswith("once:")):
                        job["next_run"] = next_run(job["spec"], now) \
                            or now + 3600
                        _atomic_write(f, job)
                        continue
                    if not sem.acquire(blocking=False):
                        continue  # at capacity — retry next scan
                    running.add(jid)

                    def _go(j=job, p=f, i=jid):
                        try:
                            fire(j, p)
                        except Exception:
                            log("fire crashed:", i)
                        finally:
                            running.discard(i)
                            sem.release()

                    threading.Thread(target=_go, daemon=True).start()
        except Exception:
            log("scan loop error")
        time.sleep(SCAN_S)


if __name__ == "__main__":
    main()
